test/extended/oauth/oauth_ldap.go

```diff
@@ -0,0 +1,128 @@
+package oauth
+
+import (
+	"encoding/base64"
+	"fmt"
+	"time"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	restclient "k8s.io/client-go/rest"
+
+	configv1 "github.com/openshift/api/config/v1"
+	osinv1 "github.com/openshift/api/osin/v1"
+	userv1client "github.com/openshift/client-go/user/clientset/versioned/typed/user/v1"
+	testutil "github.com/openshift/origin/test/extended/util"
+	oauthutil "github.com/openshift/origin/test/extended/util/oauthserver"
+)
+
+var _ = g.Describe("[Suite:openshift/oauth] LDAP IDP", func() {
+	defer g.GinkgoRecover()
+	var (
+		oc = testutil.NewCLI("oauth-ldap-idp", testutil.KubeConfigPath())
+
+		bindDN         = "cn=Manager,dc=example,dc=com"
+		bindPassword   = "admin"
+		ldapScheme     = "ldap://"
+		ldapPort       = ":389"
+		caPath         = "/var/oauth/configobjects/ldapca/ca.crt"
+		searchDN       = "ou=people,ou=rfc2307,dc=example,dc=com"
+		searchAttr     = "cn"
+		searchScope    = "one"
+		userName       = "Person1"
+		goodPass       = "foobar"
+		badPass        = "baz"
+		providerName   = "ldapidp"
+		myUserDNBase64 = base64.RawURLEncoding.EncodeToString([]byte(searchAttr + "=" + userName + "," + searchDN))
+		myUserName     = "person1smith"
+		myEmail        = "person1smith@example.com"
+	)
+
+	g.It("should authenticate against an ldap server", func() {
+		g.By("setting up an OpenLDAP server")
+		ldapService, ldapCA, err := testutil.CreateLDAPTestServer(oc)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("deploying an oauth server")
+		caConfigMap := corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "ldapca",
+			},
+			Data: map[string]string{
+				"ca.crt": string(ldapCA),
+			},
+		}
+
+		// Configure IDP
+		ldapProvider, err := oauthutil.GetRawExtensionForOsinProvider(
+			&osinv1.LDAPPasswordIdentityProvider{
+				URL:    ldapScheme + ldapService + ldapPort + fmt.Sprintf("/%s?%s?%s", searchDN, searchAttr, searchScope),
+				BindDN: bindDN,
+				BindPassword: configv1.StringSource{StringSourceSpec: configv1.StringSourceSpec{
+					Value: bindPassword,
+				}},
+				Insecure: false,
+				CA:       caPath,
+				Attributes: osinv1.LDAPAttributeMapping{
+					ID:                []string{"dn"},
+					PreferredUsername: []string{"cn"},
+					Name:              []string{"displayName"},
+					Email:             []string{"mail"},
+				},
+			},
+		)
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		// Deploy an OAuth server
+		tokenOpts, _, err := oauthutil.DeployOAuthServer(oc, []osinv1.IdentityProvider{
```

Comment:
`DeployOAuthServer` is broken since it returns a `tokenOpts` which can only be used once.  The code below works because `RequestTokenForUser` mostly ignores `tokenOpts`.

Reply:
@stlaz @sallyom ^